### PR TITLE
Enable macos packaging and tests

### PR DESCRIPTION
**Pull request by gbaraldi**
Depends on #42. I have no idea if this is working or not. This is a rebase on [this PR](https://github.com/JuliaLang/julia/pull/43529) but with a similar structure to what is on #42

### CODE REVIEW

**Comment by staticfloat** [comment]
I just merged a big re-organization, sorry about that.  You'll probably need to rebase this significantly.

**Comment by gbaraldi** [comment]
I was already building it on top of the branch of the merged PR, so it shouldn't be too hard.

**Comment by gbaraldi** [comment]
I was already building it on top of the branch of the merged PR, so it was a simple merge.

**Comment by staticfloat** [comment]
Can you enable aarch64 as well?

File: pipelines/main/platforms/build_macos.arches
Comment (staticfloat) [inline pipelines/main/platforms/build_macos.arches:1]: No need for `ARCH_ROOTFS`, `ROOTFS_TAG` or `ROOTFS_HASH` here.  Those columns can be deleted.

Comment (staticfloat) [reply pipelines/main/platforms/build_macos.arches:1]: `ARCH_ROOTFS` should only be used with sandbox; any other uses of the arch should parse it out from the triplet, [which is done by `calc_version_envs.sh`](https://github.com/JuliaCI/julia-buildkite/blob/main/utilities/calc_version_envs.sh#L4)

File: pipelines/main/platforms/build_macos.yml
Comment (staticfloat) [inline pipelines/main/platforms/build_macos.yml:13]: This is not needed; you can delete it.

File: pipelines/main/platforms/test_macos.yml
Comment (staticfloat) [inline pipelines/main/platforms/test_macos.yml:15]: This is not needed, you can delete it.

File: pipelines/main/platforms/build_macos.yml
Comment (staticfloat) [inline pipelines/main/platforms/build_macos.yml:31]: I don't think you want `gmake`.  This isn't FreeBSD, Mac's have GNU Make by default:

```
$ make --version
GNU Make 3.81
```

Comment (gbaraldi) [reply pipelines/main/platforms/build_macos.yml:31]: I tried that at first https://buildkite.com/julialang/julia-buildkite/builds/112#9229a63f-d883-4f03-8ccc-08ef8e751148 but `--output-sync` is only available on newer make. Make 3.81 is from 2006 😄 .

Comment (gbaraldi) [reply pipelines/main/platforms/build_macos.yml:31]: I guess I can just make without it, it's not that big of an issue. 
It seems like aarch64 has it and x86 doesn't for some reason.

**Comment by gbaraldi** [comment]
Yeah, I was very very confused, I was like wow, aarch64 is building very nice, then it was on the x86 builder 😄 .

**Comment by staticfloat** [comment]
ARCH_ROOTFS is supposed to refer to the architecture of the rootfs image that we're downloading, which isn't happening on these machines, so I'd rather name it just `ARCH`.

**Comment by gbaraldi** [comment]
Isn't there an ARCH variable already? That `calc_version_envs.sh` sets?

**Comment by staticfloat** [comment]
Yes, but that variable is not available at YAML parse time, sadly.

**Comment by gbaraldi** [comment]
Fair enough, I guess adding to the .arches file won't hurt. Not sure why the aarch64 build didn't work. I donwloaded it and it gave me broken dylib errors 🤔 . Will try to see if current master actually works on the m1.

**Comment by staticfloat** [comment]
The aarch64 tester is failing because the binaries are not codesigned.  I've updated our testing fork of Julia to include the latest fixes from Mosè for that, so the next build you push should have more luck there.

This is looking pretty good!  I think we're ready to embed the codesigning toolchain, so I went ahead and did that for you.  Let's have the workflow be:

* build
  - generates `.tar.gz`
* test
  - waits for build
  - downloads `.tar.gz` and extracts it
  - re-signs the binaries with ad-hoc signatures (because the build doesn't have a codesigning key, so I don't think julia will run without that)
  - runs tests
* upload
  - waits for test to complete
  - downloads `.tar.gz` and extracts it
  - re-signs the binaries with an actual codesigning key
  - generates, signs and notarizes `.dmg`
  - uploads tarball, tarball GPG key, and signed `.dmg`

**Comment by staticfloat** [comment]
The aarch64 tester is failing because the binaries are not codesigned.  I've updated our testing fork of Julia to include the latest fixes from Mosè for that, so the next build you push should have more luck there.

This is looking pretty good!  I think we're ready to embed the codesigning toolchain, so I went ahead and did that for you.  Let's have the workflow be:

* build
  - generates `.tar.gz`
* test
  - waits for build
  - downloads `.tar.gz` and extracts it
  - re-signs the binaries with ad-hoc signatures (because the build doesn't have a codesigning key, so I don't think julia will run without that)
  - runs tests
* upload
  - waits for test to complete
  - downloads `.tar.gz` and extracts it
  - re-signs the binaries with an actual codesigning key
  - generates, signs and notarizes `.dmg`
  - uploads tarball, tarball GPG key, and signed `.dmg`

**Comment by gbaraldi** [comment]
That sounds great to me, thanks :). I will see if later tonight or maybe tomorrow I can get to that. 
It's quite incredible how much faster the m1 builds, it's almost 3x faster than the normal x86.
Getting the CI for CI working here was awesome for developing this., thanks elliot and @brenhinkeller .

**Comment by staticfloat** [comment]
> It's quite incredible how much faster the m1 builds, it's almost 3x faster than the normal x86.

The x86_64 machine is pretty old: `Intel(R) Core(TM) i5-4308U CPU @ 2.80GHz`, it's a personal machine of mine that I'm using for testing.  The "true" buildbots that we tend to use are newer: `Intel(R) Core(TM) i7-8700B CPU @ 3.20GHz`, so we should get significantly better performance out of those.  Those new ones also have 12 cores instead of 4, so I think we can parallelize builds and test suites such that things run a little faster on them!  :D

File: pipelines/main/platforms/test_macos.yml
Comment (staticfloat) [inline pipelines/main/platforms/test_macos.yml:29]: If possible, let's make use of https://github.com/JuliaLang/julia/blob/master/contrib/codesign.sh here

Comment (gbaraldi) [reply pipelines/main/platforms/test_macos.yml:29]: Do we have access to it directly from the julia repo or do I add it to utilities?

**Comment by gbaraldi** [comment]
Why does adhoc signing not work here 🤔 why does apple make this so more difficult than it should be.

**Comment by DilumAluthge** [comment]
@staticfloat @gbaraldi Now that Buildkite has a built-in matrix feature, what do you guys think of trying it out for this PR?

Docs: https://buildkite.com/docs/pipelines/build-matrix

**Comment by staticfloat** [comment]
We may be able to do that, as long as we don't have to use the matrix attributes in a plugin.  (E.g. we only interpolate the values within the `command` step).

Gabriel; are you blocked on anything, or just busy?  I want to get this PR moving again, and just want to understand where things stand.

**Comment by gbaraldi** [comment]
Yeah my week was a bit busier than i expected, however today I will work on it some.

File: pipelines/main/platforms/upload_macos.yml
Comment (DilumAluthge) [inline pipelines/main/platforms/upload_macos.yml:26]:
```suggestion
    timeout_in_minutes: 90
```

Let's see if this fixes CI, and then we can figure out the interpolation later.

**Comment by gbaraldi** [comment]
That error is very informative sigh

**Comment by DilumAluthge** [comment]
It seems to be coming from here: https://github.com/buildkite/interpolate/blob/07f35b4ae2515fcb0e08e9f116bdb0efd09e44d1/parser.go#L221

File: pipelines/main/platforms/upload_macos.yml
Comment (DilumAluthge) [inline pipelines/main/platforms/upload_macos.yml:35]:
```suggestion
      DMG_NAME="$$(basename $$JULIA_BINARYDIST_FILENAME8 .tar.gz).dmg"
```

**Comment by DilumAluthge** [comment]
@gbaraldi Anything that would have been `$` in the original Bash script will need to instead be `$$` in the `.yaml` file.

File: pipelines/main/platforms/upload_macos.yml
Comment (DilumAluthge) [inline pipelines/main/platforms/upload_macos.yml:26]:
```suggestion
    timeout_in_minutes: ${TIMEOUT?}
```

File: pipelines/main/platforms/upload_macos.yml
Comment (gbaraldi) [reply pipelines/main/platforms/upload_macos.yml]: That 8 is also a typo

File: pipelines/main/platforms/upload_macos.yml
Comment (DilumAluthge) [inline pipelines/main/platforms/upload_macos.yml:35]: I suspect that `$(...)` will trip Buildkite up and that we should instead do `$(...)`.

```suggestion
      DMG_NAME="$$(basename $$JULIA_BINARYDIST_FILENAME .tar.gz).dmg"
```

Comment (gbaraldi) [reply pipelines/main/platforms/upload_macos.yml:35]: I thought the double $ was for external variables, I could be wrong however

**Comment by gbaraldi** [comment]
Thanks @DilumAluthge I was struggling with this for a while as you can see 😄 .

**Comment by DilumAluthge** [comment]
Alright, I went through and changed almost every `$` into `$$`, and now the Buildkite parsing error is gone.

File: pipelines/main/platforms/upload_macos.yml
Comment (DilumAluthge) [inline pipelines/main/platforms/upload_macos.yml:5]:
```suggestion
    # We only upload to S3 if one of the following criteria are true:
    # 1. The branch is `master`.
    # 2. The branch is `release-*`.
    # 3. The build is a tag build AND the tag is `v*`.
    # 4. The pipeline is `julia-buildkite`.
    if: (build.branch == "master") || (build.branch =~ /^release-/) || (build.tag =~ /^v/) || (pipeline.slug == "julia-buildkite")
```

File: pipelines/main/platforms/test_macos.yml
Comment (staticfloat) [inline pipelines/main/platforms/test_macos.yml:35]: Hmmm, I just saw this is in Linux as well.  Why?  @DilumAluthge

File: pipelines/main/platforms/test_macos.yml
Comment (staticfloat) [inline pipelines/main/platforms/test_macos.yml:42]: I think this most likely is not necessary on macOS, but it also doesn't hurt, I don't think.

File: pipelines/main/platforms/test_macos.yml
Comment (staticfloat) [inline pipelines/main/platforms/test_macos.yml:49]: Let's let the agent machines choose this.

**Review by staticfloat — APPROVED** [review]
Overall this looks good to me.   I think we'll clean it up as necessary, but as long as it works, I say merge when green!

File: pipelines/main/platforms/test_macos.yml
Comment (DilumAluthge) [reply pipelines/main/platforms/test_macos.yml:35]: Which line? The `OPENBLAS_NUM_THREADS`?

**Comment by DilumAluthge** [comment]
Closing/reopening doesn't retrigger Buildkite.

**Comment by gbaraldi** [comment]
Closed by mistake. I still need to add the keychain password so we correctly sign. 
Another thing, where in S3 should I upload the DMG.

**Comment by DilumAluthge** [comment]
> Another thing, where in S3 should I upload the DMG.

@staticfloat Do they just go in the same folder as the `.tar.gz` files?

File: pipelines/main/platforms/test_macos.arches
Comment (DilumAluthge) [inline pipelines/main/platforms/test_macos.arches:3]:
```suggestion
macos      aarch64-apple-darwin    true          aarch64       .              .
```

**Comment by DilumAluthge** [comment]
Why are the tests failing on Intel macOS?

**Comment by gbaraldi** [comment]
I don't know, but both x86 and arm have thousands of segfaults so i would guess that has to do with it.

**Comment by DilumAluthge** [comment]
@staticfloat It must be due to seatbelt, right?

**Comment by DilumAluthge** [comment]
If I look at the raw log, before the many segfaults start, I see this:

```
Got exception outside of a @test

  LoadError: IOError: bind: name too long (ENAMETOOLONG)

  Stacktrace:

    [1] bind(server::Sockets.PipeServer, name::String)

      @ Sockets /Users/sabae/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-solstice.0/build/default-solstice-0/julialang/julia-buildkite/julia-a5fe945b4f/share/julia/stdlib/v1.9/Sockets/src/PipeServer.jl:59
```

I'm not sure if that's related to the segfaults, or if those are two separate issues.

**Comment by gbaraldi** [comment]
I've seen those segfaults in the wild https://github.com/JuliaLang/julia/issues/44562, the socket error might be seatbelt however.

**Comment by DilumAluthge** [comment]
Have you seen the segfaults on both Intel and Apple Silicon, or just on Apple Silicon?

**Comment by gbaraldi** [comment]
I don't have an intel mac to test, but Ian had mentioned on the issue that he saw them on x86 as well.

**Comment by gbaraldi** [comment]
@staticfloat in my last version I tried moving the keychain to `Library/Keychains` it didn't work I imagine becaue the agent overrides what home is. How would I make that work?

**Comment by staticfloat** [comment]
We shouldn't need to move it to a different location.  Can you try using the full path to the keychain, even when passing to `unlock-keychain`?

**Comment by DilumAluthge** [comment]
@staticfloat Why are there four different `Macos` groups in the Buildkite UI?

**Comment by gbaraldi** [comment]
Oh yeah, macos doesn't have coreutils

**Comment by staticfloat** [comment]
Because the job that launches the macos pipelines doesn't belong to the same group as the jobs its launching.  It's one of the more frustrating limitations of buildkite's jobs, see [this thread for more](https://forum.buildkite.community/t/group-merging-doesnt-work-as-expected/1858).

**Comment by staticfloat** [comment]
@gbaraldi can you rebase on top of the latest `main` so you get the `git pager` fix?

**Comment by staticfloat** [comment]
For debugging purposes, feel free to skip the tests by having the upload not depend on the tests. :P

File: pipelines/main/platforms/upload_macos.yml
Comment (DilumAluthge) [inline pipelines/main/platforms/upload_macos.yml:14]:
```suggestion
      # - "test_${TRIPLET?}" # TODO: uncomment this line
```

**Comment by staticfloat** [comment]
I fixed the `ENAMETOOLONG`; the issue is that [libuv limits the length of unix domain socket paths](https://github.com/JuliaLang/libuv/blob/f872be3b6438b90b4690fe3ee7692c50bfd9c7c7/src/unix/pipe.c#L54-L55) to a value much smaller than the man page of `bind()` would indicate.  I've fixed this with https://github.com/JuliaCI/sandboxed-buildkite-agent/pull/9, allowing me to override the tempdir from a really long name to a much smaller prefix.

**Comment by gbaraldi** [comment]
I wonder why `security -v list-keychains -s -d user "$$KEYCHAIN_PATH"` doesn't add the keychain to the list. It works locally. Maybe it tries to write somewhere that seatbelt doesn't allow and it fails silently?

**Comment by DilumAluthge** [comment]
@staticfloat Is there a script or something that @gbaraldi can run on his local computer that will drop him into a Bash shell inside the seatbelt sandbox? That way he can reproduce the environment locally.

**Comment by staticfloat** [comment]
> @staticfloat Is there a script or something that @gbaraldi can run on his local computer that will drop him into a Bash shell inside the seatbelt sandbox? 

Yes, [the `debug_shell.jl` script](https://github.com/JuliaCI/sandboxed-buildkite-agent/blob/main/macos-seatbelt/debug_shell.jl) will do that (note you'll need to setup a `config.toml` file first, just copy the example one over to `config.toml`).  I debugged this today though, and man, it was a pain.  I eventually broke down and turned off SIP on my Macbook Pro so that I could `dtruss` the `security` tool to figure out why it wasn't adding the keychain when it should have... and it turns out it was because we were overriding `HOME`, and the `${HOME}/Library/Preferences` directory did not exist.  Le sigh.

**Comment by staticfloat** [comment]
Huzzah!  The codesigning is working!  Now to fix the `hdiutil` error and the distributed error.....

**Comment by staticfloat** [comment]
Distributed failure should be fixed by https://github.com/JuliaLang/julia/pull/44799

**Comment by gbaraldi** [comment]
You're a damn hero Elliot 🦸 . If only the command didn't fail silently, I will add a dmg upload step and I think we should be done. The DMG is also quarantined when downloaded. https://developer.apple.com/forums/thread/128166 this thread tells us to sign it too.

**Comment by gbaraldi** [comment]
It seems to work now. It has the can't open terminal bug but that's on julia itself.

**Comment by staticfloat** [comment]
The Pkg failure should be fixed by: https://github.com/JuliaLang/Pkg.jl/pull/3045

**Comment by staticfloat** [comment]
Pkg failure identified and fix pending here: https://github.com/JuliaLang/Pkg.jl/pull/3050

**Comment by staticfloat** [comment]
@gbaraldi I'm going to rebase this and start DRY'ing it a bit.  But I think we're getting about ready for a merge!

**Comment by staticfloat** [comment]
Pkg problems are worked around as of https://github.com/JuliaLang/julia/pull/44828

So we can merge this when ready.

**Comment by gbaraldi** [comment]
🥳 It's purple and green 🥳

**Comment by staticfloat** [comment]
Nice work, Gabriel!

**Pull request merged**
